File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/PolarSleepApi.swift
///  Copyright © 2024 Polar. All rights reserved.

import Foundation

/// Protocol defining methods to get Polar Sleep Data
public protocol PolarSleepApi {
    /// Get sleep recording state
    ///
    /// - Parameters:
    ///   - identifier: The Polar device ID or BT address
    ///   - timeoutMs: Timeout in milliseconds while waiting for the first sleep recording state event
    /// - Requires SDK feature(s): `PolarBleSdkFeature.feature_polar_sleep_data`
    /// - Returns: AnyPublisher emitting a single Bool value indicating if sleep recording is ongoing
    /// - Throws: ``PolarErrors/timeout(description:)`` if timeout is exceeded before state is received.
    func getSleepRecordingState(identifier: String, timeoutMs: UInt64) async throws -> Bool
    
    /// - Returns: Publisher stream of Bool values indicating if sleep recording is ongoing
    func observeSleepRecordingState(identifier: String) -> AsyncThrowingStream<[Bool], Error>
    
    /// - Returns: Publisher stream
    func stopSleepRecording(identifier: String) async throws
    
    /// - Returns: Publisher emitting an array of `PolarSleepData` for the specified period.
    func getSleep(identifier: String, fromDate: Date, toDate: Date) async throws -> [PolarSleepData.PolarSleepAnalysisResult]

    /// - Deprecated: Use ``getSleep(identifier:fromDate:toDate:)`` instead.
    @available(*, deprecated, renamed: "getSleep(identifier:fromDate:toDate:)")
    func getSleepData(identifier: String, fromDate: Date, toDate: Date) async throws -> [PolarSleepData.PolarSleepAnalysisResult]
}

public extension PolarSleepApi {
    /// Get sleep recording state with default timeout of 30 seconds.
    func getSleepRecordingState(identifier: String) async throws -> Bool {
        try await getSleepRecordingState(identifier: identifier, timeoutMs: 30_000)
    }

    @available(*, deprecated, renamed: "getSleep(identifier:fromDate:toDate:)")
    func getSleepData(identifier: String, fromDate: Date, toDate: Date) async throws -> [PolarSleepData.PolarSleepAnalysisResult] {
        return try await getSleep(identifier: identifier, fromDate: fromDate, toDate: toDate)
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarSleepApi.kt
package com.polar.sdk.api

import com.polar.sdk.api.model.sleep.PolarSleepData
import kotlinx.coroutines.flow.Flow
import java.time.LocalDate

/**
 * Polar sleep API.
 * Requires feature FEATURE_POLAR_SLEEP_DATA
 */
interface PolarSleepApi {

    /**
     * Get sleep recording state. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_SLEEP_DATA]
     *
     * @param identifier The Polar device ID or BT address
     * @return boolean value indicating if sleep recording is ongoing
     * @throws Throwable if the operation fails
     **/
    suspend fun getSleepRecordingState(identifier: String): Boolean

    /**
     * Observe sleep recording state. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_SLEEP_DATA]
     *
     * @param identifier The Polar device ID or BT address
     * @return [Flow] of boolean values indicating if sleep recording is ongoing
     */
    fun observeSleepRecordingState(identifier: String): Flow<Array<Boolean>>

    /**
     * Stop sleep recording. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_SLEEP_DATA]
     *
     * @param identifier The Polar device ID or BT address
     * @throws Throwable if sleep recording stop action cannot be sent to the device
     */
    suspend fun stopSleepRecording(identifier: String)

    /**
     * Get sleep stages and duration for a given period. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_SLEEP_DATA]
     *
     * @param identifier The Polar device ID or BT address.
     * @param fromDate The starting date of the period to retrieve sleep data from.
     * @param toDate The ending date of the period to retrieve sleep data from.
     * @return list of [PolarSleepData] representing the sleep data for the specified period.
     * @throws Throwable if the operation fails
     */
    suspend fun getSleep(identifier: String, fromDate: LocalDate, toDate: LocalDate): List<PolarSleepData>
}